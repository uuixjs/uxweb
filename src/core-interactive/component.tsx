import * as history from "history";

import {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  FocusEventHandler,
  ForwardRefRenderFunction,
  MouseEventHandler,
  ReactElement,
  ReactNode,
  Ref,
  forwardRef,
} from "react";
import {
  ensureSafeLink,
  isExternalURL,
  isMailToLink,
} from "../_utils/url-props";

import { ARIARole } from "aria-query";
import { DataTestSelectorProps } from "@uuixjs/uuixweb-lib";
import { LinkProps as RRLinkProps } from "react-router-dom";

type AnchorProps = AnchorHTMLAttributes<HTMLAnchorElement>;
type LinkProps = AnchorProps | RRLinkProps;
type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement>;
export type CoreInteractiveRenderLinkBaseProps = AnchorProps &
  Pick<CoreInteractivePublicProps, "linkTo">;

export type CoreInteractiveElement = HTMLAnchorElement | HTMLButtonElement;

export type CoreMouseEventHandlers<T = CoreInteractiveElement> = {
  /** Event called when the element is clicked. */
  onClick?: MouseEventHandler<T>;
  /** Event called when the user attempts to open a context menu. */
  onContextMenu?: MouseEventHandler<T>;
  /** Event called when when a pointing device button (e.g., a mouse's primary button) is double-clicked */
  onDoubleClick?: MouseEventHandler<T>;
  /** Event called when a pointing device (usually a mouse) is moved over the element. */
  onMouseEnter?: MouseEventHandler<T>;
  /** Event called when a pointing device (usually a mouse) is moved off the element. */
  onMouseLeave?: MouseEventHandler<T>;
};

export type CoreFocusEventHandlers<T = CoreInteractiveElement> = {
  /** Event called when the element loses focus. */
  onBlur?: FocusEventHandler<T>;
  /** Event called when the element gains focus. */
  onFocus?: FocusEventHandler<T>;
};

/**
 * CoreInteractivePublicProps is what enclosing classes should extend in order
 * to have the full compliment of props for passing through to the
 * CoreInteractive component.
 */
export interface CoreInteractivePublicProps
  extends CoreMouseEventHandlers,
    CoreFocusEventHandlers,
    DataTestSelectorProps {
  /**
   * Optional label for screen readers and assistive devices.
   */
  "aria-label"?: string;
  /**
   * If the element is a button, autofocus it when it is first rendered.
   */
  autoFocus?: boolean;
  /** Sets the element as disabled, always renders as a button. */
  disabled?: boolean;
  /** Sets the element as disabled but with no visual change, always renders as a button. */
  disabledInteraction?: boolean;
  /**
   * Sets the `download` attribute on the element. If the `linkTo` prop is set,
   * the resource will be downloaded instead of being navigated to.
   */
  download?: boolean | string;
  /** Indicates the element is a link and will navigate to the passed URL. */
  linkTo?: history.LocationDescriptor;
  /** Provides a [ref callback](https://reactjs.org/docs/refs-and-the-dom.html#callback-refs) */
  refHandler?: Ref<CoreInteractiveElement>;
  /** If the element is a link, set the value of the rel attribute */
  rel?: string;
  /** Render prop to allow for using react-router alternatives @experimental */
  renderLink?: <P extends CoreInteractiveRenderLinkBaseProps>(
    props: P
  ) => ReactNode;
  /** Replace the current entry in the history stack instead of adding a new one. */
  replace?: boolean;
  role?: ARIARole;
  /** Manually set the `tabIndex` of this element. */
  tabIndex?: number;
  /** If the element is a link, open it in a new tab. */
  targetBlank?: boolean;
  /** The elements title */
  title?: string;
}

/**
 * CoreInteractiveProps extends CoreInteractiveLinkProps, and enclosing
 * components should generate the things like className, refHandler, etc
 * internally and pass them down
 */
export interface CoreInteractiveProps extends CoreInteractivePublicProps {
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refHandler` */
  refHandler?: CoreInteractivePublicProps["refHandler"];
  className?: string;
  /**
   * The content rendered within the element.
   */
  children?: ReactNode;
}

const CoreInteractive: ForwardRefRenderFunction<
  CoreInteractiveElement,
  CoreInteractiveProps
> = (
  {
    disabled,
    // This prop does actually do something different from `disabled`, but the
    // effect is in parents like CoreButton.
    disabledInteraction,
    download,
    linkTo,
    refHandler,
    rel,
    renderLink,
    targetBlank,
    ...props
  },
  ref
) => {
  const elementRef = ref || refHandler;
  const elementProps: LinkProps | ButtonProps = props;

  // Coalesce `false` to `undefined` to prevent `disabled={false}`
  const isDisabled = disabled || disabledInteraction || undefined;

  // links
  if (linkTo && !isDisabled) {
    const anchorRef = elementRef as Ref<HTMLAnchorElement>;

    // add anchor-specific props
    const anchorProps = elementProps as LinkProps;
    if (download) {
      anchorProps.download = download;
    }
    if (rel || targetBlank) {
      anchorProps.rel = targetBlank
        ? `${rel ? rel + " " : ""}noopener noreferrer`
        : rel;
    }
    if (targetBlank) {
      anchorProps.target = "_blank";
    }

    // external link
    // if (isExternalURL(linkTo) || isMailToLink(linkTo)) {
    let href: string;
    if (typeof linkTo === "string") {
      href = linkTo;
    } else {
      // Implication: typeof props.linkTo === LocationDescriptorObject. The 'state' properties of this type cannot be included in an anchor tag.
      // We render the provided 'pathname' (which is actually a fully-qualifed URL in this case) into the anchor tag as an attempt at a gracefull fallback.
      href = (linkTo && linkTo.pathname) || "";

      // eslint-disable-next-line no-console
      console.warn(
        "External links and mailto links with LocationDescriptorObjects are not supported! Pass a string instead."
      );
    }

    // internal custom link using linkTo prop
    if (renderLink) {
      return renderLink({
        ...anchorProps,
        ref: anchorRef,
        linkTo: ensureSafeLink(linkTo),
      }) as ReactElement;
    }

    return <a {...anchorProps} href={ensureSafeLink(href)} ref={anchorRef} />;
    // }

    // internal default link
    // return (
    //   <Link {...anchorProps} to={ensureSafeLink(linkTo)} innerRef={anchorRef} />
    // );
  }

  // non-link or disabled
  return (
    <button
      {...(elementProps as ButtonProps)}
      disabled={isDisabled}
      ref={elementRef as Ref<HTMLButtonElement>}
    />
  );
};

CoreInteractive.displayName = "CoreInteractive";
const ComponentWithRef = forwardRef(CoreInteractive);
export { ComponentWithRef as CoreInteractive };
