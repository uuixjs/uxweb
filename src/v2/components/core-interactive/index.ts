export {
  CoreFocusEventHandlers,
  CoreInteractive,
  CoreInteractiveElement,
  CoreInteractiveProps,
  CoreInteractivePublicProps,
  CoreInteractiveRenderLinkBaseProps,
  CoreMouseEventHandlers,
} from "./component";
