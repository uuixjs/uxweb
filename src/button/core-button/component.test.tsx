import { mount } from "enzyme";
import { Link, MemoryRouter } from "react-router-dom";
import { setupShallowTest } from "../../../tests/helpers";
import { CoreInteractive } from "../../core-interactive";
import {
  CoreButton,
  CoreButtonComponent,
  CoreButtonProps,
  CoreButtonRounding,
  CoreButtonSize,
  CoreButtonType,
} from "./component";

function getOptionalProps(): CoreButtonProps {
  return {
    autoFocus: true,
    children: "Test Button",
    disabled: true,
    disabledInteraction: true,
    download: "Download all the things",
    fullWidth: true,
    linkTo: "/",
    onClick: jest.fn(),
    overlay: true,
    rounding: CoreButtonRounding.Left,
    size: CoreButtonSize.Large,
    tabIndex: 0,
    targetBlank: true,
    title: "Test title",
    variant: CoreButtonType.Secondary,
  };
}

const setupShallow = setupShallowTest(CoreButtonComponent);
describe("CoreButton", () => {
  it("renders CoreButton component with required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders CoreButton component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("renders CoreButton component as <a>", () => {
    const testUrl = "http://www.google.com";
    const wrapper = mount(
      <MemoryRouter>
        <CoreButton linkTo={testUrl} />
      </MemoryRouter>,
    );
    expect(wrapper.find("a").first().prop("href")).toBe(testUrl);
  });

  it("renders CoreButton component as <button> with an onclick", () => {
    const testFn = jest.fn();
    const wrapper = mount(<CoreButton onClick={testFn} />);
    wrapper.find("button").first().simulate("click");
    expect(testFn).toHaveBeenCalled();
  });

  it("renders CoreButton component as Link component", () => {
    const testPath = "/dashboard";
    const wrapper = mount(
      <MemoryRouter>
        <CoreButton linkTo={testPath} />
      </MemoryRouter>,
    );
    expect(wrapper.find(Link).first().prop("to")).toBe(testPath);
  });

  it("passes aria props to the element", () => {
    const wrapper = mount(<CoreButton aria-label="Some Action" />);
    const element = wrapper.find(CoreInteractive).first();
    expect(element.prop("aria-label")).toBe("Some Action");
  });
});
