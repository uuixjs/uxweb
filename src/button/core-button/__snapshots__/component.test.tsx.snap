// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreButton renders CoreButton component with all props 1`] = `
<Styled(Styled(Component))
  $disabled={true}
  $fullWidth={true}
  $overlay={true}
  $rounding="left"
  $size="large"
  autoFocus={true}
  disabled={true}
  disabledInteraction={true}
  download="Download all the things"
  linkTo="/"
  onClick={[MockFunction]}
  tabIndex={0}
  targetBlank={true}
  title="Test title"
>
  Test Button
</Styled(Styled(Component))>
`;

exports[`CoreButton renders CoreButton component with required props 1`] = `
<Styled(Component)
  $rounding="default"
  $size="default"
/>
`;
